### PR TITLE
[RDY] Add emberjs layer

### PR DESCRIPTION
**Pull request by robbyoconnor**
Ready for merge 

I excluded keybindings for working with Ember views since they are not used anymore.

### CODE REVIEW

File: layers/+frameworks/emberjs/packages-config.el [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+;; This file is not part of GNU Emacs.
+;;
+;;; License: GPLv3
+
+;;; Commentary:
+
+;;; Code:
+
+(defconst emberjs-packages
+  '(ember-mode)
```
Comment (AlejandroCatalina) [inline layers/+frameworks/emberjs/packages-config.el]: Isn't an endparen missing here?

**Comment by robbyoconnor** [comment]
@syl20bnr this is ready for showtime -- I'm sad I couldn't use Tomster -- their wickedly cute Hamster mascot.

**Comment by robbyoconnor** [comment]
@syl20bnr this is ready for showtime -- I'm sad I couldn't use Tomster -- their wickedly cute Hamster mascot. They control that pretty tightly :(

**Comment by robbyoconnor** [comment]
By default this uses whatever they defined -- it seems to support ido and helm as well.

File: layers/+frameworks/emberjs/README.org [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  (setq-default dotspacemacs-configuration-layers '(emberjs))
+#+end_src
+
+*** Completion System
+    By default this will use =ido=,
+    but it may be configured to use helm by adding the following in your =dotspacemacs/user-config= function:
+ #+BEGIN_SRC emacs-lisp
+  (setq ember-completion-system 'helm)
+ #+END_SRC
+
```
Comment (robbyoconnor) [inline layers/+frameworks/emberjs/README.org]: Added a note how to configure this -- I'm not going to make it work with ivy.

**Comment by robbyoconnor** [comment]
@syl20bnr hold off on this -- I wanna add [handlebars-mode](https://github.com/danielevans/handlebars-mode) -- shouldn't be long -- I promise.

**Comment by catesandrew** [comment]
nice but what about `ember-yasnippets` too? or does ember-mode already come with it? worth considering is all

**Comment by robbyoconnor** [comment]
Do you have any suggestions for keybindings? Currently it's kinda annoying `C-c .` is the prefix as it stands

**Comment by robbyoconnor** [comment]
I also added handlebars-mode to the html layer and documented it. It's used elsewhere from ember so I don't feel like isolating it is necessary.

**Comment by robbyoconnor** [comment]
@syl20bnr Ready for showtime.

File: layers/+frameworks/emberjs/README.org [hunk 3]
```diff
@@ -0,0 +1,114 @@
+#+TITLE: emberjs layer
+#+HTML_HEAD_EXTRA: <link rel="stylesheet" type="text/css" href="../css/readtheorg.css" />
```
Comment (JAremko) [inline layers/+frameworks/emberjs/README.org]: Should be  *../../../*

File: layers/+frameworks/emberjs/packages.el [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+;;; Commentary:
+
+;;; Code:
+
+(defun emberjs/init-ember-mode ()
+  (use-package ember-mode
+    :defer t))
+
+(defun emberjs/init-ember-yasnippets ()
+  (use-package ember-yansipplets
```
Comment (jaehess) [inline layers/+frameworks/emberjs/packages.el]: Possible typo:  `ember-yasnippets`?

Comment (robbyoconnor) [inline layers/+frameworks/emberjs/packages.el]: I will fix this..

**Comment by robbyoconnor** [comment]
I rebased this @syl20bnr and updated it to follow the conventions

**Comment by jaehess** [comment]
Awesome! Can't wait!

**Comment by collegeimprovements** [comment]
How can I use this layer? Can I copy the packages.el file in some custom layer ? and will it work out of the box?

**Comment by robbyoconnor** [comment]
@collegeimprovements you can grab it and clone it as a private layer.

**Comment by robbyoconnor** [comment]
But yes, grabbing it the package.el in a custom layer works.

**Comment by collegeimprovements** [comment]
Somehow for me it never initialize in ember project. I tried to install ember-mode separately in additional-packages but that also work for a single buffer only.

**Comment by robbyoconnor** [comment]
@collegeimprovements your best bet would be to cherry-pick this PR manually...and then just include it as I said. You need to read up on how private layers work.

**Comment by robbyoconnor** [comment]
@syl20bnr This was made back in February...I'm used to waiting months for my PRs but not a year!

**Comment by li-xinyang** [comment]
What's the current state of this new layer?

**Comment by robbyoconnor** [comment]
if you need it -- pull it locally and use it as a private layer. Things
can sit for awhile around here.

**Comment by robbyoconnor** [comment]
There's no remaining work on my part being done on it.

**Comment by robbyoconnor** [comment]
@syl20bnr

File: layers/+frameworks/emberjs/packages.el [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  '(ember-mode
+    ember-yasnippets))
+
+
+(defun emberjs/init-ember-mode ()
+  (use-package ember-mode
+    :defer t))
+
+(defun emberjs/init-ember-yasnippets ()
+  (use-package ember-yasipplets
```
Comment (benjamreynolds) [inline layers/+frameworks/emberjs/packages.el]: I think this needs to be `ember-yasnippets` as above.

Comment (robbyoconnor) [inline layers/+frameworks/emberjs/packages.el]: Thank you so much!

Comment (robbyoconnor) [inline layers/+frameworks/emberjs/packages.el]: Fixed.

**Comment by benjamreynolds** [comment]
Id love to see this one merged too FWIW. And I think we could do some work on spacemacs style bindings here as well. But lets see if we can get this merged first. 😉

**Comment by JAremko** [comment]
@robbyoconnor  https://travis-ci.org/syl20bnr/spacemacs/jobs/227773425 :roll_eyes:

**Comment by robbyoconnor** [comment]
I see it...I'll fix it.

**Comment by JAremko** [comment]
@robbyoconnor  :smile:

**Comment by robbyoconnor** [comment]
I don't know how to fix this...

**Comment by JAremko** [comment]
@robbyoconnor  It's just a couple of extra whitepapers here and there and a table resize. Ah also switch to `:TOC_4_gh:noexport:` from `:TOC_4_org:noexport:` EZ! 

Don't give up! :shipit:

**Comment by robbyoconnor** [comment]
:shipit: @JAremko

**Comment by JAremko** [comment]
@robbyoconnor  I knew you ca do it! :smiley:  Eventually :roll_eyes:

**Comment by robbyoconnor** [comment]
> @robbyoconnor I knew you ca do it! :smiley: Eventually :roll_eyes:

Shots fired dude.

**Comment by jsmestad** [comment]
Any update on this @syl20bnr ??

**Comment by robbyoconnor** [comment]
I fixed all CI failures.

**Comment by robbyoconnor** [comment]
Nearly 2 years later :joy:...

@syl20bnr

**Comment by robbyoconnor** [comment]
It's been 2 years since I opened this PR.. :)

**Comment by xxzozaxx** [comment]
@syl20bnr 

I think I will give you  cathedral and bazaar as a gift . if we meet.

**Comment by robbyoconnor** [comment]
@xxzozaxx Stuff happens man -- he has a family -- nobody is barred from pulling this into their local config as a private layer...

File: layers/+frameworks/emberjs/README.org [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  - Ability to build, start server, and run tests
+
+* Install
+To use this contribution add it to your =~/.spacemacs=
+
+#+begin_src emacs-lisp
+  (setq-default dotspacemacs-configuration-layers '(emberjs))
+#+end_src
+
+*** Completion System
```
Comment (JAremko) [inline layers/+frameworks/emberjs/README.org]: @robbyoconnor  This is level 1 headline not level 3 :thinking:

**Comment by syl20bnr** [comment]
Hum... thank you for still being here :-)

We need Spacemacs key bindings for this layer though.
Seems that we made a convention for these key bindings:
https://github.com/syl20bnr/spacemacs/blob/develop/doc/CONVENTIONS.org#web-frameworks

**Comment by robbyoconnor** [comment]
@syl20bnr :joy:  -- I love this project -- you'd have to REALLY do
something I didn't like for me to leave. I will try to do this before
Monday.

**Comment by robbyoconnor** [comment]
@syl20bnr I defined the keybindings.

File: layers/+frameworks/emberjs/packages.el [hunk 7]
```diff
@@ -0,0 +1,75 @@
+;;; packages.el --- emberjs layer packages file for Spacemacs.
+;;
+;; Copyright (c) 2012-2016 Sylvain Benner & Contributors
+;;
+;; Author: Robert O'Connor <robby.oconnor@gmail.com>
+;; URL: https://github.com/robbyoconnor
```
Comment (robbyoconnor) [inline layers/+frameworks/emberjs/packages.el]: I'm not sure if this is right...

Comment (sdwolfz) [reply layers/+frameworks/emberjs/packages.el]: I think the url should reflect the canonical location from where you can get the code, aka the spacemacs repo.

Comment (robbyoconnor) [reply layers/+frameworks/emberjs/packages.el]: I fixed it -- wasn't 100% sure

**Comment by sdwolfz** [comment]
@robbyoconnor do you happen to have a sample project I could try out this layer on before merging?

**Comment by robbyoconnor** [comment]
@sdwolfz I do not --Why not use the [todomvc example](https://github.com/tastejs/todomvc/tree/master/examples/emberjs)?

**Comment by sdwolfz** [comment]
Thank you :heart:!
The todomvc project gave me a bit of trouble but in the end the functionality is OK. Further improvements can be made incrementally. I reorganized the keybindings a bit in 4644a2886875f9fdf45427214def38bc99f8bbf3.

Cherry-picked into develop branch, you can safely delete your branch.

**Pull request merged**